### ISSUE TITLE
Is there a way to set the default empty element for a region?

### ISSUE THREAD

**Issue by richardprice** [GetmeUK/ContentTools#180]
So, I have several regions on a page:

1. Page title, h1
2. Section title, h2
3. Content, series of p

When I remove all content from a region, the original element within the container (in this case the h1 or h2) is removed and the region is populated by a default paragraph element - I would like to default it to the correct type for particular regions (especially if I can put the right classes on those elements as well).

Is this possible?

Cheers
Richard

**Comment by anthonyjb** [comment]
Hi @richardprice,

That's a really good question and something that we should allow to be set moving forward (so I'll update this issue to be an enhancement). As it maybe a while before I get to adding this as a core library feature one option you could take in the meantime is to patch the `_preventEmptyRegions` method for the editor app:

```javascript
// Get the editor class and patch the method that prevents empty regions
var editorCls = ContentTools.EditorApp.getCls();

editorCls.prototype._preventEmptyRegions = function() {
    var child, hasEditableChildren, i, lastModified, name, placeholder, region, type;

    // Check if any region is empty
    for (name in this._regions) {
        region = this._regions[name];
        if (region.children.length > 0) {
            continue;
        }
        
        // We have to check for elements that can receive focus as static
        // elements alone don't allow new content to be added to a region.
        lastModified = region.lastModified();
        hasEditableChildren = false;
        for (i = 0; i < region.children.length; i++) {
            child = region.children[i];
            if (child.type() !== 'Static') {
                hasEditableChildren = true;
                break;
              }
        }
        if (hasEditableChildren) {
          continue;
        }

        // If the region is empty add a placeholder element checking to see if
        // the region has specified a placeholder tag type.
        type = region.domElement().getAttribute('data-placeholder-type');

        // Insert the placeholder element for the region
        placeholder = new ContentEdit.Text(type || 'p', {}, '');
        region.attach(placeholder);

        // HACK: This action will mark the region as modified which it
        // technically isn't and so we commit the change to nullify this.
        region._modified = lastModified;
    }
};
```

In this scenario to add set a regions default automatically populate element for a region you'd do something like:

```html
<div class="[ author ]  [ editable ]" data-name="author" data-placeholder-type="h3">
    ...
</div>
```

**Comment by richardprice** [comment]
Hi @anthonyjb - thats brilliant, cheers :)  Thanks for a really good tool!

**Comment by mike00ward** [comment]
+1, this would be a really nice feature.  thanks for the patch, works great!

**Comment by anthonyjb** [comment]
I'll be pushing a fix for this in the next release (1.2.6) but just wanted to provide some detail. 

Rather than choose a specific way of defining the default placeholder element for a region I've added a method against the editor that creates them and can be overridden - IMHO this this is flexible and a solid approach to support different workflows. 

So replacing the patch above you'd now use the following code to implement the same feature:

```JavaScript
var editorCls = ContentTools.EditorApp.getCls();

editorCls.prototype.createPlaceholderElement = function(region) {
    var type = region.domElement().getAttribute('data-placeholder-type');
    return new ContentEdit.Text(type || 'p', {}, '');
};
```

The `createPlaceholderElement` is provided with the region in question as an argument allowing you to customize the placeholder element per region.
